proposed/2022/version-overrides.md

```diff
@@ -0,0 +1,119 @@
+
+The concept of dependency resolution overrides exists in many ecosystems and for NuGet, we want to ensure that there is a formal way to override package versions in the case that:
+
+- A project exists in a repository but is not part of the main set of shipping assemblies and is allowed to use newer, untested package depedencies (such as an experimental tool)
+- A project that has more advanced requirements and cannot use a single version of a package dependency for all of its target frameworks
+
+In these cases, a project can override the centrally defined package version.  However, if that project is part of a transitive graph with version conflicts, a user would still need to resolve those conflicts before restore would succeed.  The main point of this feature is to give a well documented and supported way of overriding a package version for a particular project, not to make graph resolution easier if package conflicts exist.  
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+First a user opts-in to Central Package Management with a `Directory.Packages.props` file in their directory tree.
+
+A user would specify package versions in `Directory.Packages.props`:
+
+```xml
+<Project>
+  <ItemGroup>
+    <PackageVersion Include="PackageA" Version="1.0.0" />
+    <PackageVersion Include="PackageB" Version="2.0.0" />
+  </ItemGroup>
+<Project>
+```
+
+A project would override the version with the `VersionOverride` metadata on a `<PackageReference />` item:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="PackageA" VersionOverride="3.0.0" />
+  </ItemGroup>
+<Project>
+```
+
+In this case, the project's restore graph would resolve `PackageA` to version `3.0.0`. Any project that references it would also get that version and a user would be responsible for handling an unresolved conflicts.
+
+Finally, a repo owner should be able to disable the ability for developers to override package version. This would be used for instance if someone wanted to ensure that all package versions are unified. This would be possible by setting the MSBuild property `EnablePackageVersionOverride` to `false` in a project or import like `Directory.Build.props`:
+
+```xml
+<Project>
+  <PropertyGroup>
+    <EnablePackageVersionOverride>false</EnablePackageVersionOverride>
```

Comment:
Can this be set in `Directory.Packages.props` too? This is a guess but I'm thinking it might be more intuitive to keep this setting with the overrides themselves rather than in `Directory.Build.props` which might otherwise not exist in a repo.

Reply:
As long as it's any file imported by the SDK or an explicit import, it should be fine.  

IMO it would be nice if the NuGet docs would standardize on putting all CPVM related properties in one place (i.e. Directory.Packages.props) instead of recommending some there and some in Directory.Build.props.  The two files have no relation to each other, other than that both are imported via the SDK.props, and centralizing all CPVM concerns in one place seems simpler to manage.